docs/contributor/memory_management.md

```diff
@@ -0,0 +1,168 @@
+```
+// We want to put `val` in `local0`
+local0 := {
+  tmp := incRef(val);
+  decRef(local0);
+  tmp
+}
+```
+In practice, this is facilitated by `Tuple` instructions Binaryen, similar to this:
+```
+// We want to put `val` in `local0`
+local0 := {
+  (incRef(val), decRef(local0))[0]
+}
+```
+
+**When a function returns a pointer, that pointer already has a reference count of at least one, which includes the reference from the local frame.** This means that, if the returned pointer does not escape the function, then a (net) single `decRef` should be invoked on that pointer in order to not leak memory.
+
+When a function is exiting, it is imperative to clean up all references to values in the local frame while preserving the return value. We do this by doing the following:
+
+- Calling `incRef` on the return value (which is in a local)
+- Calling `decRefIgnoreZeros` on all locals
+
+This should result in a net change of zero references on the return value, and a net change of -1 references on all other locals.
+
+**References are caller-owned.** This has a few implications:
+- If a function is called with an argument, it is the responsibility of whoever called that function to `decRef` that argument
+- If a value is passed to a function, when it returns, it will have the same reference count as when the function was called (unless it was stored in an external location, in which case the reference count may be higher).
+
+Moreover, the Grain compiler is designed to minimize the overall number of calls to `incRef` and `decRef` calls in its produced code. This means that, rather than the following (pseudocode):
+```grain
+let f = x => {
+  incRef(x)
+  foo(x)
+  decRef(x)
+  incRef(x)
+  bar(x)
+  decRef(x)
+}
+```
+Grain's compiler will emit something more like this (again, note that `x` is an argument, so it is incumbent upon `f`'s caller to `decRef` it):
+```grain
+let f = x => {
+  foo(x)
+  bar(x)
+}
+```
+
+## Disabling the Garbage Collector
+
+There are times in which the garbage collector is not desired in Grain code. The typical example of this is the implementation of the core runtime (for example the garbage collector itself), which involves working with native WebAssembly values. Consequently, Grain provides two mechanisms for disabling the garbage collector, either on a per-function or a per-file basis.
+
+First, it is possible to annotate functions with `@disableGC`. This will compile the annotated function with no garbage collection instructions. Second, it is possible to pass the `--compilation-mode=runtime` flag to the compiler (**NOTE** that this is not the only effect of this flag; see `runtime.md` for more details), which is typically done by placing this at the beginning of the file:
+```grain
+/* grainc-flags --compilation-mode=runtime */
```

Comment:
I thought runtime mode has GC but no pervasives? Though there is a `--disable-gc` flag 🤔

Reply:
Runtime mode doesn't have GC (since it implements GC). We managed to get rid of `--compilation-mode=managed-runtime` which had that behavior, since you can just use `--no-pervasives` now.

Reply:
My mistake 🤦